src/memory.rs

```diff
@@ -48,26 +38,77 @@ impl Buffer {
     /// created by `from_vec`. You may not consume a slice twice.
     /// Otherwise you risk double free panics
     pub unsafe fn consume(self) -> Vec<u8> {
-        if self.is_empty() {
-            return Vec::new();
-        }
-        let mut v = Vec::from_raw_parts(self.ptr, self.len, self.cap);
-        v.shrink_to_fit();
-        v
+        Vec::from_raw_parts(self.ptr, self.len, self.cap)
     }
 
     // this releases our memory to the caller
-    pub fn from_vec(mut v: Vec<u8>) -> Self {
+    pub fn from_vec(v: Vec<u8>) -> Self {
+        let mut v = mem::ManuallyDrop::new(v);
```

Comment:
Nice find, seems safer than mem::forget when returning.

You can now just remove `let buf =` and just directly return `Buffer {`.... good find with this wrapper

Reply:
Oh indeed. I did not see this